### ISSUE TITLE
Changing API structure for better browser and caching support

### ISSUE THREAD

**Issue by MSchmidt** [h2non/imaginary#78]
I'd like to suggest an API change that would increase support for cache proxies and general browsers.

Problems I see right now:
* if the user saves an image it's named ``crop.jpg``, ``resize.jpg`` etc. 
* cache proxies may not identify the object as image and may choose to ignore it

Both problems could be fixed by allowing images names in the URL

Old API call:
``https://imaginary.example.com/enlarge?url=https%3A%2F%2Fassets.example.com%2Fdemo.jpg&width=800&height=600``

New API call with optional file name:
``https://imaginary.example.com/enlarge/MyFunImagename.jpg?url=https%3A%2F%2Fassets.example.com%2Fdemo.jpg&width=800&height=600``

**Comment by h2non** [comment]
I'm open to dig into this. If I understand your proposal properly, the feature should easy to support.

**Comment by MSchmidt** [comment]
Yes, my first thoughts were for a completely different structure but then I ended up with this hopefully easy (and backwards-compatible) change since this is basically just allowing any additional string behind the processing endpoints.

I've just discovered this wonderful tool a few days ago so I'm not sure what your plan is. But I'm open to discuss broader ideas since I've worked with several different image processing solutions in the past.

In terms of caching the best solution is to embed all parameters in the URL path rather than parameters.

**Comment by h2non** [comment]
It's nice to read that. Backwards compatibilities is a must in `imaginary` since it's used by a bunch of people, so interface contract should be consistent at any level.

Well, I don't have big plans for `imaginary` at this time. It was designed to do one job and do it well and fast enough, and I believe it does. Most people uses it because of that, it's not hard to use or configure and it works. My thesis would be, in case that I want to radically introduce new features in `imaginary`, I would prefer to create another project, perhaps built-on-top of `imaginary`, but definitively as an isolated project.

For sure there're a lot of useful features that could be introduced as part, and cache it's just one part of the iceberg. I was thinking about this for a while and I've some interesting ideas about building a extremely versatile, modular and even hackable engine-agnostic image processing solution.
Actually, I've create a Github organization for this project called [imagestack](https://github.com/imagestack). As you can imagine, the project is pretty ambitious, but quite interesting. I'm busy with other projects at this time, but I would like to invest my time on it in the future.

Additionally, I've created another `imaginary` like package, but simpler and less versatile to cover my personal needs in a project. It's called [resizr](https://github.com/h2non/resizr). I believe it's mostly what you're looking for in order to take the advantage of your cache strategy.

I'm very open to introduce minor features in `imaginary` who doesn't implies breaking existent things, so feel free to send any proposal you've in mind. Also, if you've more ambitious ideas, we can discuss them too and perhaps place most of them in other project.

**Comment by Dynom** [comment]
Is this really the best approach? The HTTP specification specifies something to use instead, namely by sending the Content-Disposition header.

> The Content-Disposition response-header field has been proposed as a means for the origin server to suggest a default filename if the user requests that the content is saved to a file. This usage is derived from the definition of Content-Disposition in RFC 1806

https://www.w3.org/Protocols/rfc2616/rfc2616-sec19.html#sec19.5.1

It has been a while since I used that, but afaik that still works.

**Comment by h2non** [comment]
The proposal from @Dynom is definitively the best approach and more HTTP protocol friendly. Imaginary can simply build or forward that header to the client. Any other ideas? I can work on this later this week.

**Comment by Dynom** [comment]
@kirillDanshin I think this issue should be closed. This issue started with two (unrelated) questions, namely:
1. Files are downloaded without their original name, and
1. Cache proxies may not identify the object as image and may choose to ignore it

*Statement 1* could be added to the backlog as a feature that can be build if it receives enough votes. I think it's not that trivial, unless we blindly accept an argument for the filename, encode it and returning it as a `Content-Disposition` header.

*Statement 2* If a proxy doesn't respect HTTP's caching features than it's just broken and falls outside of the responsibility of Imaginary. Imaginary currently supports sending caching headers (`Cache-Control`, defined by `-http-cache-ttl`*).

@MSchmidt can you add any weight to what situation you're in that prevents proxies from caching resources?

We could, perhaps, add support for `Vary` to be more compliant with different setups and proxies, although I'm unsure if that would be really needed as it will most likely increase cache misses.

-- Just my 2 cents.

_* I just noticed that the description is wrong, the http-cache-ttl argument doesn't care about locally or remotely served files, I'll add PR for that._

**Comment by kirillDanshin** [comment]
@Dynom I think we can just support both. I like that idea.
